File: CommonPython/sort/merge_sort.py
import random

def merge_sort(arr):
	sort(arr, [0] * len(arr), 0, len(arr) - 1)
	return arr

def sort(arr, temp, l, r):
	if l >= r:
		return

	mid = (l + r) // 2
	sort(arr, temp, l, mid)
	sort(arr, temp, mid + 1, r)
	merge(arr, temp, mid + 1, l, r)

def merge(arr, temp, mid, l, r):
	for k in range(l, r + 1):
		temp[k] = arr[k]
	i = l
	j = mid
	for k in range(l, r + 1):
		if i >= mid:
			arr[k] = temp[j]
			j += 1
		elif j > r:
			arr[k] = temp[i]
			i += 1
		elif temp[i] > temp[j]:
			arr[k] = temp[j]
			j += 1
		else:
			arr[k] = temp[i]
			i += 1
	


test_array = [random.randint(0, 100) for _ in range(10)]
ans_array = sorted(test_array.copy())
merge_sort(test_array)
if test_array == ans_array:
	print("Correct")
else:
	print("Incorrect")
	print(test_array)

File: CommonJava/src/main/java/org/khwu/sort/MergeSort.java
package org.khwu.sort;

public class MergeSort implements SortStrategy {

    @Override
    public void sort(int[] arr) {
        sort(arr, new int[arr.length],0, arr.length-1);
    }

    private void sort(int[] arr, int[] aux, int lo, int hi) {
        if (lo >= hi) {
            return;
        }
        int mid = lo + (hi - lo) / 2;
        sort(arr, aux, lo, mid);
        sort(arr, aux, mid+1, hi);
        merge(arr, aux, lo, mid, hi);
    }

    private void merge(int[] arr, int[] aux, int lo, int mid, int hi) {
        for (int i = lo; i <= hi; i++) {
            aux[i] = arr[i];
        }

        int j = lo;
        int k = mid+1;
        for (int i = lo; i <= hi; i++) {
            if (j > mid) {
                arr[i] = aux[k++];
            } else if (k > hi) {
                arr[i] = aux[j++];
            } else if (aux[j] < aux[k]) {
                arr[i] = aux[j++];
            } else {
                arr[i] = aux[k++];
            }
        }
    }
}
